applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp

```diff
@@ -389,4 +396,51 @@ void ElasticIsotropic3D::CalculateCauchyGreenStrain(
     noalias(rStrainVector) = MathUtils<double>::StrainTensorToVector(E_tensor);
 }
 
+/***********************************************************************************/
+/***********************************************************************************/
+
+void ElasticIsotropic3D::AddThermalEffect(
+    ConstitutiveLaw::Parameters& rValues,
+    Vector& rStrainVector
+    )
+{
+    const auto &r_properties = rValues.GetMaterialProperties();
+    const auto &r_geometry = rValues.GetElementGeometry();
+
+    if (r_properties.Has(THERMAL_EXPANSION_COEFFICIENT)) {
+        const double alpha = r_properties[THERMAL_EXPANSION_COEFFICIENT];
+        if (r_geometry[0].SolutionStepsDataHas(TEMPERATURE)) {
+            const double gauss_point_temperature = ConstitutiveLawUtilities<VoigtSize>::
```

Comment:
Note that doing it in this way implies looping the nodes twice (one for the `TEMPERATURE` and one for `REFERENCE_TEMPERATURE`. I'd rather remove the  `GetNodalVariableOnIntegrationPoint`.

Reply:
mm here I'm computing the `TEMPERATURE` and the `REFERENCE_TEMPERATURE` at the gauss point (interpolation from the nodes). What do you mean by removing the `GetNodalVariableOnIntegrationPoint`? I think I need both. :)

Reply:
You need both, but for each call of the `GetNodalVariableOnIntegrationPoint` you are looping the nodes. You can loop the nodes once and access to the `TEMPERATURE` and `REFERENCE_TEMPERATURE`.